### ISSUE TITLE
Those meteor-ios support authentication via facebook and twitter?

### ISSUE THREAD

**Issue by emmanuelbuah** [martijnwalraven/meteor-ios#10]
If so, I/the community will greatly appreciate a code sample. Thanks for the awesome work on the lib.

**Comment by martijnwalraven** [comment]
I'm caught up in a project and haven't been able to work on authentication. But it is still on my to do list.

**Comment by emmanuelbuah** [comment]
Thanks @martijnwalraven and sorry for the late response. I'm going to be using/digging into meteor-ios in next couple of days. Let me know if you want me to submit a pull request for accounts (fb/twitter) integration or if you want to take a stub at what needs to change first.

**Comment by martijnwalraven** [comment]
Well, I think I'd like to take some time to make sure we'd be doing accounts integration right and secure. It would be good to hear about your experiences though, and let me know if you have any questions about the existing code. I should have more time to work on this project in the near future.

**Comment by emmanuelbuah** [comment]
Will do.

**Comment by emmanuelbuah** [comment]
@martijnwalraven I finally got to investigate how oauth authentication works in meteo. Meteor.loginWith<Service> on meteor works in 2 phases. 

### Phase 1: 
1.1 Authenticate with service X yielding a serviceData (ex. https://github.com/meteor/meteor/blob/release/METEOR%401.0.3.2/packages/github/github_server.js)
1.2 ServiceData is saved in pendingCredential collection in combination with a random gen credentials token and secrete. This is used in phase 2. 
1.3 Calls Meteor method  "login" with credentials token and secrete 

### Phase 2:
2.1 Login method retrieves credentials record from pendingCredentials collection, matches on serviceData.id to users.services."type".id and returns { type: "servicetype', userId: <matched-userid>}
2.2 Attempts login (internal validation + call to Accounts.validateLoginAttempt), if user doesn't exist, it create the user. If it exists, it updates the service record with the new accesstoken. Finally it deletes the record in pendingCredential collection
2.3 Generates an authenticated token for user and returns it to the client. This token is used in subsequent re-authentication via resume service.

Since Facebook ios sdk handles all of phase one (resulting in an accesstoken at the end of the process) except for 1.3. Integration with meteor-ios will look like this:

1. Process authentication via facebook ios sdk
2. Call custom Meteor method "savePendingCredentialsForAccessToken(service,accesstoken)" which returns { credentialToken: x, credentialSecrete: y}. savePendingCredentialsForAccessToken will prep serviceData and save it to pendingCredential collection. 
3. Call Meteor method "login" to complete the process, which will return { id: <userid>, token: <authtoken>, tokenExpires: <token expiration date>} 

What are your thoughts and how best to go about changes.

**Comment by smothers** [comment]
Here's how I went about solving Facebook SDK integration meteor/meteor#4441. It's the same method used in ObjectiveDDP. I've done zero auditing of the security of this approach and it seems like it will be a while before there is official support for OAuth via 3rd party SDKs in Meteor.

**Comment by emmanuelbuah** [comment]
@smothers do you have a working sample with all the changes/modification?

**Comment by emmanuelbuah** [comment]
@smothers looks like fb-accouns-ddp- https://github.com/jasper-lu/facebook-ddp holds everything. thanks. I will check it out.

**Comment by smothers** [comment]
@emmanuelbuah yes, here's the addition that accepts an accessToken https://github.com/jasper-lu/facebook-ddp/blob/master/facebook_server.js#L52

It does not use facebook's [appsecret_proof ](https://developers.facebook.com/docs/facebook-login/access-tokens#architecture) to validate the request, but it does check the app id of the response against the app id on the server.

**Comment by emmanuelbuah** [comment]
So my guess is to make similar modification for other oauth services (twitter, github etc) that needs native support

**Comment by smothers** [comment]
That looks like the easiest path for now.

Alternatively you could try to make this change after ddp login is called, but before pending_credentials is queried. This way you could bypass the need to make an initial http request to add the token to pending_credentials.

**Comment by emmanuelbuah** [comment]
@smothers I agree. I think this is the easiest way. I another is to introduce a ddp method for each service to make pending_credentials entry before calling login so we don't have to mess with core accounts-facebook, ...

**Comment by bitterloa** [comment]
Hi @smothers  and @emmanuelbuah I wanted to check in and see how things were going regarding Facebook/Oauth for you? I've read over all of the documentation linked to in this thread and still absorbing it all -- and my basic question is, in your opinion is using ddp/meteor-ios "production ready"? You both seem well versed in the coding, I'm playing catch-up and am truly loving the Meteor platform and XCode7/Swift2. I'm just trying to figure out if I can reliably add production-ready code for an iOS native app using Meteor as a backend, that requires FB and/or Twitter/Oauth logins. I don't mind digging in and getting a little dirty with custom code but at the end of the day would love to know your thoughts on: are you using this in production apps, and if so once setup is it a pretty solid solution (until Meteor implements something better perhaps?) Thanks fellas

**Comment by epaga** [comment]
So @martijnwalraven @emmanuelbuah @smothers etc...at this point, what would you all say is the best practice for logging in via Facebook/Twitter OAuth when using meteor-ios in an app?

**Comment by tonyxiao** [comment]
@epaga implement a custom login handler for meteor. Like so

```coffeescript
Accounts.registerLoginHandler 'facebook', (serviceData) ->
  loginRequest = serviceData['facebook']

  # if 'fb-access' is not present, then this is a different type of login. Ignore.
  if !loginRequest?
    return undefined

  accessToken = loginRequest.accessToken
  if !accessToken?
    error = new Meteor.Error(401, 'Facebook login without accessToken')
    logger.error(error)

  response = FacebookLogin.instance().handle(accessToken, this.connection.id)
```

**Comment by epaga** [comment]
Thanks @tonyxiao - sorry if I'm being dense, but what would the meteor-ios side of this look like?

**Comment by tonyxiao** [comment]
@epaga 
```swift
    func loginWithFacebook(#accessToken: String, expiresAt: NSDate) -> RACSignal {
        return meteor.loginWithMethod(method: "login", params: [[
            "fb-access": [
                "accessToken": accessToken,
                "expireAt": expiresAt.timeIntervalSince1970
            ]
        ]])
    }
```

**Comment by epaga** [comment]
OK I can get the access token with the Facebook/Twitter SDKs, but for the life of me I'm not figuring out the server side of things. You reference "FacebookLogin", is this pseudo-code or is there some package I don't know about? Does this create a Meteor user?

The meteor packages "twitter", "accounts-twitter", etc., all seem to be working from the browser use case - they have things like "_storePendingCredential" for multiple browser tabs, etc. Surely there is a simple way to create / login a Meteor user server-side using the access token (& secret?) created with the client-side mobile SDKs...?

**Comment by tonyxiao** [comment]
@epaga This is a piece of custom login handler. Hook into meteor is here

`Accounts.registerLoginHandler 'facebook', (serviceData) ->`

This is part of the meteor API (those not really documented well). It expects an object return value with the key `userId` present if login is successful. 

And the `FacebookLogin` part is custom logic. For us it looks something like this. 

```coffeescript
  _fetchFromFacebook: (accessToken) ->
    try
      return Meteor.http.call('GET',
        "https://graph.facebook.com/me?access_token=#{accessToken}").data
    catch error
      error = new Meteor.Error(500, 'RuntimeException from Facebook Auth', error.message)
      logger.error(error)
      throw error

  handle: (accessToken, connectionId) ->
    userInfo = @_fetchFromFacebook(accessToken)

    if !userInfo?
      error = new Meteor.Error(500, 'Failed to link Facebook', accessToken)
      logger.error(error)
      throw error

    userInfo.accessToken = accessToken

    existingUser = Users.findOne
      { 'services.facebook.id': userInfo.id }

    if existingUser?
      # logging in should set the active timestamp on the user
      Users.update existingUser._id,
        $set: lastActive: new Date()

      return {
        userId: existingUser._id,
      }

    userId = Users.insert
      lastActive: new Date()
      services:
        facebook: userInfo

    user = Users.findOne(userId)

    return {
      userId: user._id,
    }

```
